File: tessoku/chapter9/a66.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type UnionFind struct {
	Parents    []int
	GroupSizes []int
}

// NewUnionFind is UnionFind構造体の初期化
// Parentsはrootがない-1を設定
// GroupSizesは、グループの要素数は初期値1
func NewUnionFind(n int) *UnionFind {
	p := make([]int, n+1)
	g := make([]int, n+1)
	for i := 1; i <= n; i++ {
		p[i] = -1
		g[i] = 1
	}

	return &UnionFind{
		Parents:    p,
		GroupSizes: g,
	}
}

// Root is xの頂点を返す
func (uf *UnionFind) Root(x int) int {
	for {
		if uf.Parents[x] == -1 {
			break
		}
		x = uf.Parents[x]
	}
	return x
}

// unite is 要素uとvの統合
func (uf *UnionFind) unite(u, v int) {
	uRoot := uf.Root(u)
	vRoot := uf.Root(v)
	if uRoot == vRoot {
		return // 同じグループなら結合不要
	}

	if uf.GroupSizes[uRoot] < uf.GroupSizes[vRoot] {
		uf.Parents[uRoot] = vRoot
		uf.GroupSizes[vRoot] += uf.GroupSizes[uRoot]
	} else {
		uf.Parents[vRoot] = uRoot
		uf.GroupSizes[uRoot] += uf.GroupSizes[vRoot]
	}
}

// same is 要素uとvが同じグループか判定する
func (uf *UnionFind) same(u, v int) bool {
	if uf.Root(u) == uf.Root(v) {
		return true
	} else {
		return false
	}
}

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, q int
	fmt.Fscan(in, &n, &q)

	uf := NewUnionFind(n)

	for i := 0; i < q; i++ {
		var a, b, c int
		fmt.Fscan(in, &a, &b, &c)

		switch a {
		case 1:
			uf.unite(b, c)
		case 2:
			if uf.same(b, c) {
				fmt.Fprintln(out, "Yes")
			} else {
				fmt.Fprintln(out, "No")
			}
		}
	}
}

File: tessoku/chapter9/a66.rb
class UnionFind
  attr_accessor :parents, :group_sizes

  def initialize(n)
    @parents = Array.new(n + 1) { -1 } # 親要素管理配列の初期化
    @group_sizes = Array.new(n + 1) { 1 } # グループ内の要素サイズの初期化
  end

  # 要素xが所属するtreeのrootを返す
  # @param [Integer] x
  # @return [Integer]
  def root(x)
    loop do
      break if @parents[x] == -1
      x = @parents[x]
    end

    x
  end

  # 頂点uと頂点vを統合する
  # @param [Integer] u
  # @param [Integer] v
  # @return [void]
  def unite(u, v)
    u_root = root(u)
    v_root = root(v)

    return if u_root == v_root # すでに同じグループなら統合不要

    # 頂点数が大きい方を上にして結合する
    if @group_sizes[u_root] < @group_sizes[v_root]
      @parents[u_root] = v_root
      @group_sizes[v_root] += @group_sizes[u_root]
    else
      @parents[v_root] = u_root
      @group_sizes[u_root] += @group_sizes[v_root]
    end
  end

  # 頂点uと頂点vが同じグループか判定する
  # @param [Integer] u
  # @param [Integer] v
  # @return [Boolean]
  def same(u, v)
    root(u) == root(v)
  end
end

n, q = gets.chomp.split(" ").map(&:to_i)
uf = UnionFind.new(n)

q.times do
  q = gets.chomp.split(" ").map(&:to_i)
  case q[0]
  when 1
    uf.unite(q[1], q[2])
  when 2
    puts uf.same(q[1], q[2]) ? "Yes" : "No"
  end
end